crates/ntx-builder/src/state/mod.rs

```diff
@@ -170,7 +173,28 @@ impl State {
                 continue;
             }
 
-            let notes = account.notes().take(limit.get()).cloned().collect::<Vec<_>>();
+            // Remove notes that have failed too many times.
+            account.drain_failing_notes(Self::MAX_NOTE_ATTEMPTS);
+
+            // Select notes from the account that can be consumed or are ready for a retry.
+            let notes = account
+                .notes()
+                .take(limit.get())
+                .filter(|&note| {
+                    // Filter based on execution hint.
+                    let block_num = self.chain_tip_header.block_num();
+                    let can_consume = note
+                        .to_inner()
+                        .metadata()
+                        .execution_hint()
+                        .can_be_consumed(block_num)
+                        .unwrap_or(true);
+                    // Filter based on attempts also.
+                    can_consume
+                        && Self::backoff_has_passed(note.last_attempt(), block_num, note.attempts())
```

Comment:
Should we merge this all into something like

```rust
fn is_note_available(&self, note: &InflightNetworkNote) -> bool {
    // Check execution hint.
    ...
    // Check backoff
}
```

This does mess with your tests because creating notes with those conditionals might be a pain. But it might be worth trying it out with some basic `P2ID` or other simple note constructors. @PhilippGackstatter where should we go looking for a real but simple note builder? We essentially only care about manipulating the `execution_hint`.

Reply:
Alternate idea: move the `can_consume` check as a method into the `InflightNetworkNote`. This might be sensible to do regardless; I don't mind where these single use functions live tbh, but in terms of ownership it probably makes sense that its there..

Or even move the entire `is_note_available` into the note impl since all it really requires from the state is the current chain tip.